PostgreSQL reload virtual columns on update via RETURNING clause

abaldwin88 opened the pull request:
Extends Active Record to automatically reload virtual columns on update when using PostgreSQL. This is done by issuing a single UPDATE query that includes a RETURNING clause.

### Motivation

Saves an extra round trip to the database for reload and removes a developer pitfall around stale values.

### Detail

Given a `Post` model represented by the following schema:
```ruby
create_table :posts do |t|
  t.integer :upvotes_count
  t.integer :downvotes_count
  t.virtual :total_votes_count, type: :integer, as: "upvotes_count + downvotes_count", stored: true
end
```
`total_votes_count` will reflect the sum of upvotes and downvotes after `update` is successfully called. Prior to this change calling `reload` would have been necessary to obtain the value calculated by the database.
```ruby
post = Post.find(1)
post.update(upvotes_count: 2, downvotes_count: 2)
# Calling `post.reload` no longer necessary
post.total_votes => 4
```

* At this moment MySQL and SQLite adapters do not support this behavior

### Additional information

* Follow-up from PR #48241
* Related to PR #48434
* Closes #48423 when using Postgresql

CC: @nvasilevski

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

ideasasylum commented:
I just checked out Rails main to see how the generated columns work with `previous_changes`. It worked on `create` but doesn't on `update` 😞 

Then I tried this branch and it works as expected 🥳 

Given a virtual column like:

```ruby
add_column :products, :premium, :virtual, type: :boolean, as: "price >= 1000", stored: true
```

I could do:

```ruby
product = Product.create! user: u, price: 100
product.premium
=> false
product.update price: 2000
=> true
product.premium
=> true
product.previous_changes
=> 
{"price"=>[100, 2000],
 "updated_at"=>[Sun, 23 Jul 2023 22:22:35.416328000 UTC +00:00, Sun, 23 Jul 2023 22:22:51.945466000 UTC +00:00],
 "premium"=>[false, true]}
```

Yay! I can see the changes caused by the update of the `premium` virtual column

I hope this makes it into 7.1 as the virtual columns will solve a gnarly problem for me if the dirty tracking works for them

excid3 commented:
Ran into this one today as well. Would love to see this fixed.

abaldwin88 commented:
Rebased onto the latest from main. Added a test for dirty tracking behavior based on the example provided by @ideasasylum

nvasilevski commented on activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:
```diff
@@ -195,9 +196,9 @@ def insert(arel, name = nil, pk = nil, id_value = nil, sequence_name = nil, bind
       alias create insert
 
       # Executes the update statement and returns the number of rows affected.
```
If I'm following the changes correctly the biggest concern with this PR is going to be the changed signature of `update` and `exec_update` methods. It used to return the number of rows affected but now it suddenly returns a rich result object. Returning the result may be a better design overall but it can't happen without a deprecation. 
Based on the changes we made to `exec_insert` we have the following options:
- Return rich result but only if `returning` argument is non-nil
- new method that always returns rich result which will be used by `_update_record` 
I don't feel completely comfortable of the same method returning different types of values depending on arguments passing but I imagine a new method will raise even more concerns so I think returning result only if `returning` is present is a better option

abaldwin88 replied:
Yes that's definitely a concern and I wasn't sure how the rails team would prefer to approach it.

> Return rich result but only if returning argument is non-nil

This sounds like a pragmatic compromise to me.

nvasilevski commented on activerecord/lib/active_record/connection_adapters/mysql2/database_statements.rb:
```diff
@@ -48,7 +48,24 @@ def exec_delete(sql, name = nil, binds = []) # :nodoc:
-        alias :exec_update :exec_delete
+
+        def exec_update(sql, name = nil, binds = [], _returning = nil) # :nodoc:
+          affected_rows = if without_prepared_statement?(binds)
+            with_raw_connection do |conn|
+              @affected_rows_before_warnings = nil
+              execute_and_free(sql, name) do
+                @affected_rows_before_warnings || conn.affected_rows
+              end
+            end
+          else
+            exec_stmt_and_free(sql, name, binds) do |stmt|
+              stmt.affected_rows
+            end
+          end
```
This is essentially a copy of `exec_delete`, right? 
So technically the method could be written as:

```suggestion
     exec_delete(...)
     build_result(columns: [], rows: [], affected_rows: affected_rows)
```

Though it does look weird so I think instead the logic from `exec_delete` should be extracted into a new method like:

```ruby
def _exec_statement(sql, name = nil, binds = [], _returning = nil)
   <same body as current version of `exec_delete`>
end

alias :exec_delete :_exec_statement

def exec_update(sql, name = nil, binds = [], _returning = nil) # :nodoc:
    affected_rows = exec_statement(...)
    build_result(columns: [], rows: [], affected_rows: affected_rows)
end
```

abaldwin88 replied:
Yes, it is a copy. I wasn't sure how closely the Rails codebase followed the rule of three. So, I elected to open the PR without an abstraction and see if there was feedback.

I like the suggestion as written and will implement it as such.

nvasilevski reviewed (commented):
I think the biggest issue with the current implementation is changing return value of public methods which is a breaking change and normally requires a deprecation cycle

Other than that I think it's a very valuable feature to be merged and also a safe one since most of the changes happen behind the scenes and don't require any public APIs to be exposed

abaldwin88 commented:
Thanks for the feedback @nvasilevski. I've implemented your suggestion for keeping the signature the same expect when `returning` is specified. I also made a couple other small revisions based on your input, added docs, and rebased onto latest from `main`.

Build error from Rubocop appears to be from `main`. This PR does not modify `action_mailer_basics.md`.

rafaelfranca commented on activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb:
```diff
@@ -91,6 +96,16 @@ def sql_for_insert(sql, pk, binds, returning) # :nodoc:
+          return super unless returning_columns&.any?
+
+          returning_columns_statement = returning_columns.map { |c| quote_column_name(c) }.join(", ")
+          sql = "#{sql} RETURNING #{returning_columns_statement}"
```
Instead of concat sqls, can we just build the right AST in the right place? We probably need to build the AST before passing to `exec_update` in the `update` method.

abaldwin88 replied:
Sorry I'm not entirely clear on the ask here. Is there something different here than `sql_for_insert` which uses the same code pattern?

I'm planning to move this method from being postgres specific and instead put it under the abstract class.

abaldwin88 replied:
My two cents is any further changes around this would be better handled as a separate follow-up PR? A refactor really ought to also include changes to `sql_for_insert`. However, including changes to that method here would be too broad and make this PR unfocused.

quentindemetz replied:
@rafaelfranca I don't think Arel supports `Returning` nodes yet (implemented in https://github.com/rails/rails/pull/47161 🙏🏻), also https://github.com/rails/rails/pull/48241 was merged in 2023 with the same strategy 💡

rafaelfranca commented on activerecord/lib/active_record/connection_adapters/sqlite3/database_statements.rb:
```diff
@@ -56,11 +56,15 @@ def internal_exec_query(sql, name = nil, binds = [], prepare: false, async: fals
         end
 
-        def exec_delete(sql, name = "SQL", binds = []) # :nodoc:
+        def _exec_statement(sql, name = "SQL", binds = []) # :nodoc:
```
Do we need to introduce this new method? We could rename `exec_delete` to `exec_statement` and remove all `exec_delete` references.

abaldwin88 replied:
After the rebase this change will no longer be necessary.

Latest WIP: https://github.com/rails/rails/compare/main...abaldwin88:wip_prototype_rebase_virtual_stored_column_returning_on_update

rafaelfranca commented on activerecord/lib/active_record/persistence.rb:
```diff
@@ -1208,10 +1208,21 @@ def _touch_row(attribute_names, time)
+
+      return result if result.is_a?(Integer)
+
+      returning_columns.zip(result.rows.first).each do |column, value|
```
Can we avoid calling `result.rows.first` twice?

rafaelfranca commented on activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:
```diff
@@ -195,9 +196,20 @@ def insert(arel, name = nil, pk = nil, id_value = nil, sequence_name = nil, bind
-        exec_update(sql, name, binds)
+        result = exec_update(sql, name, binds, returning: returning)
+
+        if result.is_a?(ActiveRecord::Result) && returning.nil?
```
I don't think I understood this conditional. It is that sometime we return `nil` and sometimes we return a result? If that is the case I'd prefer to keep `update` returning where it was and introduce another method to return the result for the adapter that needs it.

rafaelfranca reviewed (commented):
This PR needs a rebase

abaldwin88 committed:
PostgreSQL reload virtual columns on update via RETURNING clause

abaldwin88 commented on activerecord/test/schema/sqlite_specific_schema.rb:
```diff
@@ -2,7 +2,8 @@
   create_table :defaults, force: true do |t|
-      t.integer :random_number, default: -> { "ABS(RANDOM())" }
+      t.virtual :virtual_stored_number, type: :integer, as: "random_number * 10", stored: true if supports_virtual_columns?
+      t.integer :random_number, default: -> { "ABS(RANDOM() % 100)" }
```
This way it will match the behavior of `random_number` in the postgresql schema

abaldwin88 commented on activerecord/test/schema/sqlite_specific_schema.rb:
```diff
@@ -2,7 +2,8 @@
 ActiveRecord::Schema.define do
   create_table :defaults, force: true do |t|
-      t.integer :random_number, default: -> { "ABS(RANDOM())" }
+      t.virtual :virtual_stored_number, type: :integer, as: "random_number * 10", stored: true if supports_virtual_columns?
```
Ensures certain code paths / edge cases will be exercised during tests

abaldwin88 reviewed (commented):
Rebased and addressed feedback.

With the latest changes in `main` I see a pretty clear path to adding this functionality for SQLite and Maria as well. That can be done in a follow-up PR.

afn commented:
Is this PR mergable or is there any outstanding feedback that needs to be addressed? Happy to help with anything to get this moved along.

samenick commented:
Hey, any updates with this PR?

rafaelfranca committed:
Deprecate `ActiveRecord::ConnectionAdapters::Column#auto_populated?` in favor of `auto_populated_on_insert?`
